horovod/torch/__init__.py

```diff
@@ -227,6 +227,22 @@ def broadcast_optimizer_state(optimizer, root_rank):
     callbacks = {}
     occurrences = collections.defaultdict(int)
 
+    # Returns the full type structure of the possibly nested objects for recursive casting back
+    def _get_types(x):
+        if isinstance(x, collections.Iterable):
+            return type(x), [_get_types(xi) for xi in x]
+        else:
+            return type(x)
+
+    # Casts an object encoded in a tensor back into its original type and subtypes
+    def _recursive_cast(x, dtype):
+        if isinstance(dtype, tuple):
+            t, dtypes = dtype
+            x = t(x)
+            return t([_recursive_cast(x[i], dtypes[i]) for i in range(len(x))])
```

Comment:
Does this work with both lists, set, and dictionaries?

Reply:
It will work with lists, sets, and tuples.  However, it will not work with dictionaries as is, but I can make that change.

Reply:
Let's do it - maybe there will be dictionaries, too.

Reply:
The one catch here is that dictionaries cannot be converted to tensors using `torch.Tensor`.  So in order to support dictionaries, we'd need to make changes to our encoding as well as our decoding.

Reply:
Oh, I see.  I guess we could wait until this becomes necessary then.

Reply:
Sounds good.